numpyro/contrib/autoguide/__init__.py

```diff
@@ -142,31 +108,31 @@ class AutoContinuous(AutoGuide):
         Blei
 
     :param callable model: A Pyro model.
-    :param callable init_loc_fn: A per-site initialization function.
+    :param callable init_strategy: A per-site initialization function.
         See :ref:`autoguide-initialization` section for available functions.
     """
-    def __init__(self, rng, model, get_params_fn, prefix="auto", init_loc_fn=init_to_median):
-        # Wrap model in a `substitute` handler to initialize from `init_loc_fn`.
-        # Use `block` to not record sample primitives in `init_loc_fn`.
-        model = substitute(model, substitute_fn=block(seed(init_loc_fn, rng)))
+    def __init__(self, rng, model, get_params_fn, prefix="auto", init_strategy=init_to_median):
+        self.rng = rng
+        self.init_strategy = init_strategy
+        model = seed(model, rng)
         super(AutoContinuous, self).__init__(model, get_params_fn, prefix=prefix)
 
     def _setup_prototype(self, *args, **kwargs):
         super(AutoContinuous, self)._setup_prototype(*args, **kwargs)
+        # FIXME: without block statement, get AssertionError: all sites must have unique names
+        init_params, is_valid = block(find_valid_initial_params)(self.rng, self.model, *args,
+                                                                 init_strategy=self.init_strategy,
+                                                                 **kwargs)
         self._inv_transforms = {}
-        unconstrained_sites = {}
         for name, site in self.prototype_trace.items():
             if site['type'] == 'sample' and not site['is_observed']:
                 # TODO: handle dynamic support
```

Comment:
@neerajprad I intended to handle dynamic support here but the situation is more complicated, which requires we change the logic of SVI. SVI also have constrain_fn, which is static after `svi_init` method. So I would like to defer it to another PR. Maybe we should do the same thing as in HMC for SVI (which is to use `log_density` with `skip_dist_transforms=True`). Here inv_transforms still refers to transform to `site['fn'].support`, rather than `site['fn'].base_dist.support`.

Reply:
Hmm, in SVI, we initialize `param` handler by its init value, rather than from `init_strategy`, so I think we need to handle dynamic for SVI first, then look into this PR later.